lib/ABI/FunctionType.cpp

```diff
@@ -52,74 +55,95 @@ bool verify(const SortedVector<RegisterType> &UsedRegisters,
   return true;
 }
 
-constexpr static model::PrimitiveTypeKind::Values
-selectTypeKind(model::Register::Values) {
-  // TODO: implement a way to determine the register type. At the very least
-  // we should be able to differentiate GPRs from the vector registers.
-
-  return model::PrimitiveTypeKind::PointerOrNumber;
+/// \brief A helper used to differentiate vector registers.
+///
+/// \param Register Any CPU register the model is aware of.
+///
+/// \return `true` if `Register` is a vector register, `false` otherwise.
+constexpr bool isVectorRegisterHelper(model::Register::Values Register) {
+  namespace PrimitiveTypeKind = model::PrimitiveTypeKind;
+  return model::Register::primitiveKind(Register) == PrimitiveTypeKind::Float;
 }
 
-static model::QualifiedType
-buildType(model::Register::Values Register, model::Binary &TheBinary) {
-  model::PrimitiveTypeKind::Values Kind = selectTypeKind(Register);
-  size_t Size = model::Register::getSize(Register);
-  return model::QualifiedType(TheBinary.getPrimitiveType(Kind, Size), {});
-}
-
-static model::QualifiedType
-buildGenericType(model::Register::Values Register, model::Binary &TheBinary) {
-  constexpr auto Kind = model::PrimitiveTypeKind::Generic;
-  size_t Size = model::Register::getSize(Register);
-  return model::QualifiedType(TheBinary.getPrimitiveType(Kind, Size), {});
-}
+/// \brief Helps detecting unsupported ABI trait definition with respect to
+///        the way they return the return values.
```

Comment:
Would it be correct to say that this is an ABI validation routine?
Would it make sense to `static_assert` this for all of the ABIs?

Reply:
This is exactly what happens: https://github.com/revng/revng/pull/232/files/be4b91c10ae98f9d98a2523024c149135ea81392..3633f0aec491208fd97fcd22d4c6c175e08676eb#diff-d3267e3664e7756da13306db15374fd5236b1a440f47ef16c0a0d9878d62a8a5R990

Reply:
(The link doesn't seem to work properly, so I'll just paste a bit of code instead):
```cpp
      static_assert(verifyReturnValueLocationHelper<ABI>(),
                    "ABIs where non-first argument GPR is used for passing the "
                    "address of the space allocated for big return values are "
                    "not currently supported.");
```

Reply:
Done.